cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java

```diff
@@ -227,40 +229,41 @@ public String getDisplayName() {
-        try {
-            CpsThreadGroup g = getProgramPromise().get();
-            return getThread(g);
-        } catch (ExecutionException e) {
-            throw new IOException(e);
-        }
+        return getThread(getThreadGroupSynchronously());
     }
 
-    private static final ExecutorService getProgramPromiseExecutorService = new ContextResettingExecutorService(Executors.newCachedThreadPool(new NamingThreadFactory(new DaemonThreadFactory(), "CpsStepContext.getProgramPromise")));
-    private @Nonnull ListenableFuture<CpsThreadGroup> getProgramPromise() {
-        if (programPromise == null) {
-            final SettableFuture<CpsThreadGroup> f = SettableFuture.create();
-            // TODO is there some more convenient way of writing this using Futures.transform? FlowExecutionOwner.get should really return ListenableFuture<FlowExecution>
-            getProgramPromiseExecutorService.submit(new Runnable() {
-                @Override public void run() {
-                    try {
-                        ListenableFuture<CpsThreadGroup> pp;
-                        CpsFlowExecution flowExecution = getFlowExecution();
-                        while ((pp = flowExecution.programPromise) == null) {
-                            Thread.sleep(100); // TODO why is this occasionally not set right away?
-                        }
-                        f.set(pp.get());
-                    } catch (Throwable x) { // from getFlowExecution() or get()
-                        f.setException(x);
-                    }
-                }
-            });
-            programPromise = f;
+    private @Nonnull CpsThreadGroup getThreadGroupSynchronously() throws InterruptedException, IOException {
+        if (threadGroup == null) {
+            ListenableFuture<CpsThreadGroup> pp;
+            CpsFlowExecution flowExecution = getFlowExecution();
+            while ((pp = flowExecution.programPromise) == null) {
+                Thread.sleep(100); // TODO why is this occasionally not set right away?
+            }
+            try {
+                threadGroup = pp.get();
+            } catch (ExecutionException e) {
+                throw new IOException(e);
+            }
         }
-        return programPromise;
+        return threadGroup;
     }
 
+    // As in c16a522, using jenkins.util.Timer for this could deadlock. TODO would like a standard unbounded executor service.
+    private static final ExecutorService isReadyExecutorService = new ContextResettingExecutorService(Executors.newCachedThreadPool(new NamingThreadFactory(new DaemonThreadFactory(), "CpsStepContext.isReady")));
+    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_BAD_PRACTICE")
     @Override public boolean isReady() {
-        return getProgramPromise().isDone();
+        if (threadGroup == null) {
+            // but start computing it
+            isReadyExecutorService.submit(new Callable<Void>() {
```

Comment:
> Could still have multiple submitted concurrently here, yes?

Yes. (There could _not_ have been before. As noted in the PR description, I am _not_ bothering to avoid duplicate such tasks since calls to `isReady` are not that common anyway.)



> Did this just move the original problem?

The original problem was something else, that `getThreadSynchronously` was waiting forever on a task. Now it does not wait on any task, so that problem cannot recur.